pgrx/src/datum/date.rs

```diff
@@ -64,6 +47,32 @@ impl From<TimestampWithTimeZone> for Date {
     }
 }
 
+/// Create a [`Date`] from a [`pg_sys::DateADT`]
+///
+/// Note that [`pg_sys::DateADT`] is an `i32` as a Julian day offset from the "Postgres epoch".
+///
+/// The details of the encoding may also prove surprising, for instance:
+/// - It is not a Gregorian calendar date, but rather a Julian day
+/// - Despite having the numerical range for it, it does not support values before Julian day 0
+///   (4713 BC, January 1), nor does it support many values far into the future.
+/// - There is no "year zero" in either the Julian or the Gregorian calendars,
+///   so you may have to account for a "skip" from 1 BC to 1 AD
+/// - Some values such as `i32::MIN` and `i32::MAX` have special meanings as infinities
+impl TryFrom<pg_sys::DateADT> for Date {
+    type Error = DateTimeConversionError;
+    #[inline]
+    fn try_from(value: pg_sys::DateADT) -> Result<Self, Self::Error> {
+        if (i32::MIN + 1..-2451545).contains(&value) {
+            Err(DateTimeConversionError::FieldOverflow) // but it's an underflow...
+                                                        // well, underflow isn't right either
+        } else if ((2147483494 - 2451545)..i32::MAX).contains(&value) {
+            Err(DateTimeConversionError::FieldOverflow)
```

Comment:
These need proper variants.

Reply:
This is important for the correctness of the changes in this test, so I can't skimp on it before landing this.

Reply:
`DATE_END_JULIAN` and `POSTGRES_EPOCH_JDATE`, so yeah.

Reply:
Added `DatetimeConversionError::OutOfRange` to address this.